In [1]:
import pandas as pd
from pydataset import data

initiated datasets repo at: /Users/ryanmiller-alt/.pydataset/


## Exercise 1 - Copy the code from the lesson to create a dataframe full of student grades.

In [4]:
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})

type(df)

pandas.core.frame.DataFrame

1a. Create a column named passing_english that indicates whether each student has a passing grade in english.

In [7]:
df['passing_english'] = df.english > 70
df

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
11     Alan    92       62       72            False

1b. Sort the english grades by the passing_english column. How are duplicates handled?

In [18]:
df.sort_values(by='passing_english')

name  math  english  reading  passing_english
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True

In [ ]:
# Before sorting by the actual english grades, the duplicates are sorted by index order.

1c. Sort the english grades first by passing_english and then by student name. All the students that are failing english should be first, and within the students that are failing english they should be ordered alphabetically. The same should be true for the students passing english. (Hint: you can pass a list to the .sort_values method)

In [99]:
df.sort_values(by = ['passing_english','name'])

name  math  english  reading  passing_english  overall_grade
11     Alan    92       62       72            False      75.333333
8    Albert    92       62       87            False      80.333333
7     Marie    93       63       90            False      82.000000
6    Thomas    82       64       81            False      75.666667
4       Ada    77       92       98             True      89.000000
3     Billy    98       96       88             True      94.000000
10    Isaac    92       99       93             True      94.666667
1      Jane    88       79       67             True      78.000000
5      John    79       76       93             True      82.666667
9   Richard    69       80       94             True      81.000000
0     Sally    62       85       80             True      75.666667
2     Suzie    94       74       95             True      87.666667

1d. Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.

In [101]:
df.sort_values(by=['passing_english', 'english'], ascending = [False,False])

name  math  english  reading  passing_english  overall_grade
10    Isaac    92       99       93             True      94.666667
3     Billy    98       96       88             True      94.000000
4       Ada    77       92       98             True      89.000000
0     Sally    62       85       80             True      75.666667
9   Richard    69       80       94             True      81.000000
1      Jane    88       79       67             True      78.000000
5      John    79       76       93             True      82.666667
2     Suzie    94       74       95             True      87.666667
6    Thomas    82       64       81            False      75.666667
7     Marie    93       63       90            False      82.000000
8    Albert    92       62       87            False      80.333333
11     Alan    92       62       72            False      75.333333

1e. Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.

In [23]:
df['overall_grade'] = (df.english + df.math + df.reading)/3
df

name  math  english  reading  passing_english  overall_grade
0     Sally    62       85       80             True      75.666667
1      Jane    88       79       67             True      78.000000
2     Suzie    94       74       95             True      87.666667
3     Billy    98       96       88             True      94.000000
4       Ada    77       92       98             True      89.000000
5      John    79       76       93             True      82.666667
6    Thomas    82       64       81            False      75.666667
7     Marie    93       63       90            False      82.000000
8    Albert    92       62       87            False      80.333333
9   Richard    69       80       94             True      81.000000
10    Isaac    92       99       93             True      94.666667
11     Alan    92       62       72            False      75.333333

## Exercise 2 - Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:

In [26]:
mpg = data('mpg')
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

2a. How many rows and columns are there?

ANS: 234 rows x 11 columns

2b. What are the data types of each column?

In [32]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


2c. Summarize the dataframe with .info and .describe

In [34]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

2d. Rename the cty column to city.

In [37]:
mpg = mpg.rename(columns={'cty':'city'})
mpg

manufacturer   model  displ  year  cyl       trans drv  city  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16   26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

2e. Rename the hwy column to highway.

In [40]:
mpg = mpg.rename(columns = {'hwy' : 'highway'})
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

2f. Do any cars have better city mileage than highway mileage?

In [41]:
mpg[mpg.city > mpg.highway]

Empty DataFrame
Columns: [manufacturer, model, displ, year, cyl, trans, drv, city, highway, fl, class]
Index: []

2g. Create a column named mileage_difference this column should contain the difference between highway and city mileage for each car.

In [74]:
mpg['mileage_difference'] = mpg.highway - mpg.city
mpg = mpg.drop(columns = ['mileage difference', 'average mileage'])
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  average_mileage  
1    compact                  11             23.5  
2    compact                   8             25.0  
3    compact                  11             25.5  
4    compact                   9             25.5  
5    compact                  10             21.0  
..       ...                 ...              ...  
230  midsize                   9             23.5  
231  midsize                   8             25.0  
232  midsize                  10             21.0  
233  midsize                   8             22.0  
234  midsize                   9             21.5  

[234 rows x 13 columns]

2h. Which car (or cars) has the highest mileage difference?

In [75]:
mpg.sort_values(by = 'mileage_difference', ascending = False).head(5)

manufacturer       model  displ  year  cyl     trans drv  city  highway  \
107        honda       civic    1.8  2008    4  auto(l5)   f    24       36   
223   volkswagen  new beetle    1.9  1999    4  auto(l4)   f    29       41   
1           audi          a4    1.8  1999    4  auto(l5)   f    18       29   
229   volkswagen      passat    1.8  1999    4  auto(l5)   f    18       29   
36     chevrolet      malibu    3.5  2008    6  auto(l4)   f    18       29   

    fl       class  mileage_difference  average_mileage  
107  c  subcompact                  12             30.0  
223  d  subcompact                  12             35.0  
1    p     compact                  11             23.5  
229  p     midsize                  11             23.5  
36   r     midsize                  11             23.5

In [ ]:
# The 2008 Honda Civic and 1999 Volkswagon Beetle.

2i. Which compact class car has the lowest highway mileage? The best?

In [76]:
mpg[mpg['class'] == 'compact'].sort_values(by = 'highway').head(2)

manufacturer  model  displ  year  cyl       trans drv  city  highway fl  \
220   volkswagen  jetta    2.8  1999    6    auto(l4)   f    16       23  r   
221   volkswagen  jetta    2.8  1999    6  manual(m5)   f    17       24  r   

       class  mileage_difference  average_mileage  
220  compact                   7             19.5  
221  compact                   7             20.5

In [77]:
mpg[mpg['class'] == 'compact'].sort_values(by = 'highway', ascending = False).head(2)

manufacturer    model  displ  year  cyl       trans drv  city  highway fl  \
213   volkswagen    jetta    1.9  1999    4  manual(m5)   f    33       44  d   
197       toyota  corolla    1.8  2008    4  manual(m5)   f    28       37  r   

       class  mileage_difference  average_mileage  
213  compact                  11             38.5  
197  compact                   9             32.5

In [ ]:
# The 1999 Volkswagon Jetta has both the highest and lowest highway mileage for a car depending on the displacement.

2j. Create a column named average_mileage that is the mean of the city and highway mileage.

In [78]:
mpg['average_mileage'] = (mpg.highway + mpg.city)/2
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  average_mileage  
1    compact                  11             23.5  
2    compact                   8             25.0  
3    compact                  11             25.5  
4    compact                   9             25.5  
5    compact                  10             21.0  
..       ...                 ...              ...  
230  midsize                   9             23.5  
231  midsize                   8             25.0  
232  midsize                  10             21.0  
233  midsize                   8             22.0  
234  midsize                   9             21.5  

[234 rows x 13 columns]

2k. Which dodge car has the best average mileage? The worst?

In [79]:
mpg[mpg.manufacturer == 'dodge'].sort_values(by = "average_mileage").head(5)

manufacturer                model  displ  year  cyl       trans drv  city  \
70        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
66        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
60        dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
55        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
74        dodge  ram 1500 pickup 4wd    5.9  1999    8    auto(l4)   4    11   

    highway fl   class  mileage_difference  average_mileage  
70       12  e  pickup                   3             10.5  
66       12  e  pickup                   3             10.5  
60       12  e     suv                   3             10.5  
55       12  e  pickup                   3             10.5  
74       15  r  pickup                   4             13.0

In [80]:
mpg[mpg.manufacturer == 'dodge'].sort_values(by = "average_mileage", ascending = False).head(2)

manufacturer        model  displ  year  cyl     trans drv  city  highway  \
38        dodge  caravan 2wd    2.4  1999    4  auto(l3)   f    18       24   
42        dodge  caravan 2wd    3.3  2008    6  auto(l4)   f    17       24   

   fl    class  mileage_difference  average_mileage  
38  r  minivan                   6             21.0  
42  r  minivan                   7             20.5

In [ ]:
# See above for answers!

## Exercise 3 - Load the Mammals dataset. Read the documentation for it, and use the data to answer these questions:

In [86]:
Mammals = data('Mammals')
data('Mammals', show_doc = True)

Mammals

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Garland(1983) Data on Running Speed of Mammals

### Description

Observations on the maximal running speed of mammal species and their body
mass.

### Usage

    data(Mammals)

### Format

A data frame with 107 observations on the following 4 variables.

weight

Body mass in Kg for "typical adult sizes"

speed

Maximal running speed (fastest sprint velocity on record)

hoppers

logical variable indicating animals that ambulate by hopping, e.g. kangaroos

specials

logical variable indicating special animals with "lifestyles in which speed
does not figure as an important factor": Hippopotamus, raccoon (Procyon),
badger (Meles), coati (Nasua), skunk (Mephitis), man (Homo), porcupine
(Erithizon), oppossum (didelphis), and sloth (Bradypus)

### Details

Used by Chappell (1989) and Koenker, Ng and Portnoy (1994) to illustrate the
fitting of piecewise linear curves.

### Source

Garland, T. (

3a-c. How many rows and columns are there?
 What are the data types?
 Summarize the dataframe with .info and .describe

In [87]:
Mammals.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 107 entries, 1 to 107
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   weight    107 non-null    float64
 1   speed     107 non-null    float64
 2   hoppers   107 non-null    bool   
 3   specials  107 non-null    bool   
dtypes: bool(2), float64(2)
memory usage: 2.7 KB


In [88]:
Mammals.describe()

weight       speed
count   107.000000  107.000000
mean    278.688178   46.208411
std     839.608269   26.716778
min       0.016000    1.600000
25%       1.700000   22.500000
50%      34.000000   48.000000
75%     142.500000   65.000000
max    6000.000000  110.000000

In [90]:
Mammals.head(10)

weight  speed  hoppers  specials
1   6000.0   35.0    False     False
2   4000.0   26.0    False     False
3   3000.0   25.0    False     False
4   1400.0   45.0    False     False
5    400.0   70.0    False     False
6    350.0   70.0    False     False
7    300.0   64.0    False     False
8    260.0   70.0    False     False
9    250.0   40.0    False     False
10  3800.0   25.0    False      True

In [ ]:
# 107 rows, 4 columns

3d. What is the the weight of the fastest animal?

In [92]:
Mammals.sort_values(by = 'speed', ascending = False).head(5)

weight  speed  hoppers  specials
53    55.0  110.0    False     False
39    37.0  105.0    False     False
35    50.0  100.0    False     False
41    34.0   97.0    False     False
42    30.0   97.0    False     False

In [ ]:
# Weight of fastest mammal is 55.0.

In [111]:
Mammals[Mammals.speed == Mammals.speed.max()].weight

53    55.0
Name: weight, dtype: float64

3e. What is the overal percentage of specials?

In [132]:
'{:.2%}'.format(Mammals.specials.mean())

'0.93%'

3f. How many animals are hoppers that are above the median speed? What percentage is this?

In [118]:
(Mammals[(Mammals.speed > Mammals.speed.median()) & (Mammals.hoppers == True)].count()).speed

7

In [131]:
"{:.2%}".format(((Mammals[(Mammals.speed > Mammals.speed.median()) & (Mammals.hoppers == True)].count())/Mammals.count()).speed)

'6.54%'